File: f#/src/SodiumFRP.FSharp/_Root/LazyBehavior.fs
namespace SodiumFRP.FSharp

type internal ValuePropertyOrLazyInitialValue<'T> = | ValueProperty of 'T | LazyInitialValue of Lazy<'T>

type internal LazyBehavior<'T> =
    inherit Behavior<'T>
    
    internal new ((transaction : Transaction), stream, lazyInitialValue) =
        let mutable valuePropertyOrLazyInitialValue = LazyInitialValue lazyInitialValue
        let setValueProperty value = valuePropertyOrLazyInitialValue <- ValueProperty value
        let sampleNoTransaction () =
            match valuePropertyOrLazyInitialValue with
                | ValueProperty value -> value
                | LazyInitialValue l ->
                    let value = l.Value
                    valuePropertyOrLazyInitialValue <- ValueProperty <| value
                    value
        let ensureValueIsCreated () = sampleNoTransaction () |> ignore
        { inherit Behavior<_>(stream, setValueProperty, sampleNoTransaction) }
        then transaction.Sample ensureValueIsCreated

File: dotnet/src/SodiumFRP/Sodium/LazyBehavior.cs
using System;

namespace Sodium
{
    public class LazyBehavior<T> : Behavior<T>
    {
        internal Lazy<T> LazyInitialValue;

        internal LazyBehavior(Stream<T> stream)
            : base(stream, default(T))
        {
        }

        internal LazyBehavior(Transaction trans, Stream<T> stream, Lazy<T> lazyInitialValue)
            : base(stream, default(T))
        {
            this.LazyInitialValue = lazyInitialValue;

            trans.Sample(this.EnsureValueIsCreated);
        }

        protected override void NotUsingInitialValue()
        {
            base.NotUsingInitialValue();

            this.LazyInitialValue = null;
        }

        internal override T SampleNoTransaction()
        {
            this.EnsureValueIsCreated();

            return this.ValueProperty;
        }

        private void EnsureValueIsCreated()
        {
            if (this.UsingInitialValue && this.LazyInitialValue != null)
            {
                this.ValueProperty = this.LazyInitialValue.Value;
                this.LazyInitialValue = null;
            }
        }
    }
}